core/foundation/inc/ROOT/RRangeCast.hxx

```diff
@@ -0,0 +1,207 @@
+constexpr bool hasBeginEnd(...)
+{
+   return false;
+}
+
+template <typename T, typename WrappedIterator_t, bool isDynamic>
+class TypedIter {
+
+public:
+   TypedIter(WrappedIterator_t const &iter) : fIter{iter} {}
+
+   TypedIter &operator++()
+   {
+      ++fIter;
+      return *this;
+   }
+   TypedIter operator++(int)
+   {
+      TypedIter tmp(*this);
+      operator++();
+      return tmp;
+   }
+   bool operator==(const TypedIter &rhs) const { return fIter == rhs.fIter; }
+   bool operator!=(const TypedIter &rhs) const { return fIter != rhs.fIter; }
+
+   void swap(TypedIter &other) { fIter.swap(other.fIter); }
+
+   // We want to know at compile time whether dynamic_cast or static_cast is
+   // used. First of all to avoid overhead, but also to avoid a compiler
+   // error when using dynamic_cast on a non-polymorphic class. In C++17,
+   // this can be done easily with `if constexpr`, but for the older
+   // standards we have to use a more verbose alternative. Both ways are
+   // explicitely implemented for different standards, so that when the
+   // minimum C++ standard for ROOT is raised to C++17 it's easy to remember
+   // that we can avoid much boilerplate code in this file.
+#if (__cplusplus < 201700L)
+   T operator*() { return ROOT::Internal::RCast<T, isDynamic>::cast(*fIter); }
+#else
+   T operator*()
+   {
+      if constexpr (isDynamic) {
+         return dynamic_cast<T>(*fIter);
+      } else {
+         if constexpr (std::is_polymorphic<RBaseType<T>>::value) {
+            assert(dynamic_cast<T>(*fIter));
+         }
+         return static_cast<T>(*fIter);
+      }
+   }
+#endif
+
+private:
+   WrappedIterator_t fIter;
+};
+
+} // namespace Internal
+
+/// Wraps any collection that can be used in range-based loops and applies
+/// `static_cast<T>` or `dynamic_cast<T>` to each element.
+template <typename T, bool isDynamic, typename Range_t>
+class RRangeCast {
+
+public:
+   RRangeCast(Range_t &&inputRange) : fInputRange{inputRange}
+   {
+      static_assert(ROOT::Internal::hasBeginEnd<Range_t>(0),
+                    "Type with no `begin` or `end` method passed to `RRangeCast`");
+   }
+
+   using const_iterator = Internal::TypedIter<T, decltype(std::cbegin(std::declval<Range_t>())), isDynamic>;
+   const_iterator begin() const { return std::cbegin(fInputRange); }
+   const_iterator end() const { return std::cend(fInputRange); }
+
+   using iterator = Internal::TypedIter<T, decltype(std::begin(std::declval<Range_t>())), isDynamic>;
+   iterator begin() { return std::begin(fInputRange); }
+   iterator end() { return std::end(fInputRange); }
+
+private:
+   Range_t fInputRange;
+};
+
+/// Takes any collection that can be used in range-based loops and applies
+/// static_cast<T> to each element. This function can be used for example to
+/// cast all objects in a RooAbsCollection when iterating over them.
+/// Example:
+/// ~~~{.cpp}
+/// for(auto *pdf : ROOT::RangeStaticCast<RooAbsPdf const*>(argSet)) {
```

Comment:
Consider using a non-RooFit example for general ROOT usage.

Reply:
Okay!